### PR TITLE
ADM-1000 [frontend][backend]: support lead time for changes for single data source in the metrics page

### PR DESCRIPTION
**Pull request by zhou-yinyuan**
## Summary

...

## Before

_Description_

**Screenshots**
If applicable, add screenshots to help explain behavior of your code.

## After

_Description_

**Screenshots**
If applicable, add screenshots to help explain behavior of your code.

## Note

_Null_

### CODE REVIEW

File: frontend/src/containers/MetricsStep/SouceControlConfiguration/index.tsx [hunk 1]
```diff
@@ -83,6 +85,8 @@ export const SourceControlConfiguration = () => {
     totalSourceControlNumber !== 0 &&
     loadingCompletedNumber === totalSourceControlNumber;
 
+  const isShowRemoveButton = pipelineTools.type === PIPELINE_TOOL_OTHER_OPTION ? totalSourceControlNumber > 1 : true;
```
Comment (davidwangcn9) [inline frontend/src/containers/MetricsStep/SouceControlConfiguration/index.tsx:88]: Do we need the logic that only support remove when totalSourceControlNumber > 1?
How about always allow to remove for source control

File: frontend/src/containers/MetricsStep/DeploymentFrequencySettings/SingleSelection/index.tsx [hunk 2]
```diff
@@ -69,8 +69,8 @@ export const SingleSelection = ({
         <Autocomplete
           disableClearable
           data-test-id={labelId}
-          options={sortDisabledOptions(settings, options)}
-          getOptionDisabled={(option: string) => getDisabledOptions(settings, option)}
+          options={sortDisabledOptions(settings, options, deploymentFrequencySettings)}
```
Comment (davidwangcn9) [inline frontend/src/containers/MetricsStep/DeploymentFrequencySettings/SingleSelection/index.tsx:72]: Why it need deploymentFrequencySettings?

File: frontend/src/containers/MetricsStep/index.tsx [hunk 3]
```diff
@@ -185,22 +185,23 @@ const MetricsStep = () => {
...
+              </MetricsSelectionTitle>
+              <DeploymentFrequencySettings />
+            </MetricSelectionWrapper>
+          )}
+          <MetricSelectionWrapper aria-label='Source Control Configuration Section'>
+            <MetricsSelectionTitle aria-label='Source Control configuration title'>
+              Source control configuration {pipelineTools.type !== PIPELINE_TOOL_OTHER_OPTION && '(optional)'}
             </MetricsSelectionTitle>
-            <DeploymentFrequencySettings />
+            <SourceControlConfiguration />
```
Comment (davidwangcn9) [inline frontend/src/containers/MetricsStep/index.tsx:202]: Why DeploymentFrequencySettings?

**Review by davidwangcn9 — APPROVED** [review]

File: backend/src/test/java/heartbeat/controller/pipeline/dto/response/PipelineTransformerTest.java [hunk 4]
```diff
@@ -25,14 +25,39 @@ void shouldTransformToPipeline() {
...
-		assertEquals(pipeline.id, "SLUG");
-		assertEquals(pipeline.name, "Name");
-		assertEquals(pipeline.repository, "repository");
-		assertEquals(pipeline.orgId, orgId);
-		assertEquals(pipeline.orgName, orgName);
-		assertEquals(pipeline.steps.size(), 1);
-		assertEquals(pipeline.steps.get(0), "Name1");
+		assertEquals("SLUG", pipeline.id);
+		assertEquals("Name", pipeline.name);
+		assertEquals("repository", pipeline.repository);
```
Comment (guzhongren) [inline backend/src/test/java/heartbeat/controller/pipeline/dto/response/PipelineTransformerTest.java:30]: assertion is the same as line 33

File: backend/src/main/java/heartbeat/controller/pipeline/dto/response/PipelineTransformer.java [hunk 5]
```diff
@@ -2,18 +2,23 @@
...
+		String repoName = GithubUtil.getGithubUrlFullName(repositoryFullName);
+
 		return Pipeline.builder()
 			.orgId(orgId)
 			.orgName(orgName)
 			.id(dto.getSlug())
 			.name(dto.getName())
-			.repository(dto.getRepository())
+			.repository(repositoryFullName)
+			.repoName(repoName)
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/controller/pipeline/dto/response/PipelineTransformer.java:21]: so confused field `repository` and `repoName`

File: backend/src/test/java/heartbeat/controller/pipeline/dto/response/PipelineTransformerTest.java [hunk 6]
```diff
@@ -25,14 +25,39 @@ void shouldTransformToPipeline() {
...
+			.name("Name")
+			.repository("organization/repository")
+			.steps(steps)
+			.build();
+
+		Pipeline pipeline = PipelineTransformer.fromBuildKitePipelineDto(dto, orgId, orgName);
 
+		assertEquals("SLUG", pipeline.id);
+		assertEquals("Name", pipeline.name);
+		assertEquals("organization/repository", pipeline.repository);
```
Comment (guzhongren) [inline backend/src/test/java/heartbeat/controller/pipeline/dto/response/PipelineTransformerTest.java:55]: same as line 58

File: frontend/__tests__/containers/MetricsStep/PipelineConfiguration/PipelineConfiguration.test.tsx [hunk 7]
```diff
@@ -153,7 +149,7 @@ describe('DeploymentFrequencySettings', () => {
 
     await userEvent.click(getByTestId('AddIcon'));
 
-    expect(addADeploymentFrequencySetting).toHaveBeenCalledTimes(1);
+    expect(addAPipelineSetting).toHaveBeenCalledTimes(1);
```
Comment (guzhongren) [inline frontend/__tests__/containers/MetricsStep/PipelineConfiguration/PipelineConfiguration.test.tsx:152]: PLZ assert the parameters of the `addAPipelineSetting`

File: frontend/__tests__/containers/MetricsStep/PipelineConfiguration/PipelineConfiguration.test.tsx [hunk 8]
```diff
@@ -162,10 +158,66 @@ describe('DeploymentFrequencySettings', () => {
     await act(async () => {
       await userEvent.click(getAllByRole('button', { name: REMOVE_BUTTON })[0]);
     });
-    expect(deleteADeploymentFrequencySetting).toHaveBeenCalledTimes(1);
+    expect(deleteAPipelineSetting).toHaveBeenCalledTimes(1);
```
Comment (guzhongren) [inline frontend/__tests__/containers/MetricsStep/PipelineConfiguration/PipelineConfiguration.test.tsx:161]: A blank line between g w t

File: frontend/__tests__/containers/MetricsStep/PipelineConfiguration/PipelineConfiguration.test.tsx [hunk 9]
```diff
@@ -162,10 +158,66 @@ describe('DeploymentFrequencySettings', () => {
...
+  it('should call updateDeploymentFrequencySetting function and clearErrorMessages function when select organization and the value doesnt match pipeline info', async () => {
+    const { getAllByRole, getByRole } = setup();
+
+    await act(async () => {
+      await userEvent.click(getAllByRole('button', { name: LIST_OPEN })[0]);
+    });
+    const listBox = within(getByRole('listbox'));
+    await act(async () => {
+      await userEvent.click(listBox.getByText('mockOrgName2'));
+    });
```
Comment (guzhongren) [inline frontend/__tests__/containers/MetricsStep/PipelineConfiguration/PipelineConfiguration.test.tsx:173]: userEvent.click 已经被act包了一层了，你确定还要再被包一层？

**Review by guzhongren — COMMENTED** [review]
some comments

File: frontend/__tests__/containers/MetricsStep/PipelineConfiguration/PipelineConfiguration.test.tsx (hunk 9 above)
Comment (zhou-yinyuan) [reply frontend/__tests__/containers/MetricsStep/PipelineConfiguration/PipelineConfiguration.test.tsx:173]: done

File: frontend/__tests__/containers/MetricsStep/PipelineConfiguration/PipelineConfiguration.test.tsx (hunk 8 above)
Comment (zhou-yinyuan) [reply frontend/__tests__/containers/MetricsStep/PipelineConfiguration/PipelineConfiguration.test.tsx:161]: done

File: frontend/__tests__/containers/MetricsStep/PipelineConfiguration/PipelineConfiguration.test.tsx (hunk 7 above)
Comment (zhou-yinyuan) [reply frontend/__tests__/containers/MetricsStep/PipelineConfiguration/PipelineConfiguration.test.tsx:152]: this function is no any parameter

File: backend/src/test/java/heartbeat/controller/pipeline/dto/response/PipelineTransformerTest.java (hunk 6 above)
Comment (zhou-yinyuan) [reply backend/src/test/java/heartbeat/controller/pipeline/dto/response/PipelineTransformerTest.java:55]: It's not same, and repository is ssh or http url.  Now I have modified the test to mock the GithubUtil static method

File: backend/src/main/java/heartbeat/controller/pipeline/dto/response/PipelineTransformer.java (hunk 5 above)
Comment (zhou-yinyuan) [reply backend/src/main/java/heartbeat/controller/pipeline/dto/response/PipelineTransformer.java:21]: It's not same, and repository is ssh or http url, but repo name is repo name. For example, repository is `git@github.com:au-heartbeat/Heartbeat.git` but repoName is `au-heartbeat/Heartbeat`

File: backend/src/test/java/heartbeat/controller/pipeline/dto/response/PipelineTransformerTest.java (hunk 4 above)
Comment (zhou-yinyuan) [reply backend/src/test/java/heartbeat/controller/pipeline/dto/response/PipelineTransformerTest.java:30]: done

**Pull request merged**